File: commands/probability/howlucky.ts
import { ChatInputCommandInteraction, EmbedBuilder, SlashCommandSubcommandBuilder } from 'discord.js';
import Bot from '../../bot/bot';
import SlashCommand from '../../bot/slashcommand';
import { verifyProbability } from './howmanyruns';

const choose = (n: number, k: number) => {
  if (k < 0 || k > n) return 0;
  if (k > n - k) k = n - k;
  let result = 1;
  for (let i = 1; i <= k; i++) result *= (n - k + i) / i;
  return Math.round(result);
};

export default class HowLucky extends SlashCommand {
  constructor(client: Bot) {
    const data = new SlashCommandSubcommandBuilder()
      .setName('howlucky')
      .setDescription('Calculate how lucky a drop was')
      .addStringOption(option => option
        .setName('probability')
        .setDescription('x in 1/x')
        .setRequired(true)
      )
      .addIntegerOption(option => option
        .setName('try-count')
        .setDescription('The number of items/tries taken')
        .setMinValue(1)
        .setRequired(true)
      )
      .addIntegerOption(option => option
        .setName('success-count')
        .setDescription('The number of successes/rare items dropped')
        .setMinValue(1)
      );
    super(data, client, { belongsTo: 'probability' });
  }

  async execute(interaction: ChatInputCommandInteraction): Promise<void> {
    const probabilityInput = interaction.options.getString('probability')!;
    const tryCount = interaction.options.getInteger('try-count')!;
    const successCount = interaction.options.getInteger('success-count') ?? 1;
    
    const probability = verifyProbability(probabilityInput);

    if (probability === -1) {
      const embed = new EmbedBuilder()
        .setColor(this.client.config.colors.RED)
        .setTitle('Invalid Format')
        .setDescription('The argument \`probability\` must be a decimal number or a fraction, and be between 0 and 1.');
      interaction.reply({ embeds: [embed], ephemeral: true });
      return;
    }

    // P(X >= m | n tries) =
    // 1 - P(X < m | n tries) =
    // 1 - \sum_{k = 0}^{m - 1} {n \choose k} p^k (1 - p)^{n - k}
    //
    // Here m = successCount; n = tryCount

    let result = 1;
    for (let k = 0; k < successCount; k++) {
      result -= choose(tryCount, k) * Math.pow(probability, k) * Math.pow(1 - probability, tryCount - k);
    }

    const displayProbability = (result * 100).toPrecision(result >= 0.01 ? 3 : 2);

    const embed = new EmbedBuilder()
      .setColor(this.client.config.colors.BLUE)
      .setTitle('How lucky were you?')
      .setDescription(`With a drop chance of **${probabilityInput}**, getting ${successCount > 1 ? `**${successCount}** drops` : 'a drop'} within **${tryCount}** tries\nhas a **${displayProbability}%** chance of happening.`);

    interaction.reply({ embeds: [embed] });    
  }
}

File: src/commands/probability/howlucky.rs
use probability::distribution::{Binomial, Distribution};
use serenity::all::{CommandInteraction, CommandOptionType};
use serenity::async_trait;
use serenity::builder::{CreateCommandOption, CreateEmbed};
use serenity::client::Context;

use super::howmanyruns::verify_probability;
use crate::color::Colors;
use crate::commands::SlashSubCommand;
use crate::database::Database;
use crate::interaction::{BetterResponse, InteractionCustomGet};
use crate::Result;

#[derive(Default)]
pub struct HowLucky;

#[async_trait]
impl SlashSubCommand for HowLucky {
  fn register(&self) -> CreateCommandOption {
    CreateCommandOption::new(
      CommandOptionType::SubCommand,
      "howlucky",
      "Calculate how lucky a drop was"
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::String,
        "probability",
        "A decimal number or fraction between 0 and 1 (not inclusive)"
      )
      .required(true)
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::Integer,
        "try-count",
        "The number of items/tries taken"
      )
      .min_int_value(1)
      .required(true)
    )
    .add_sub_option(
      CreateCommandOption::new(
        CommandOptionType::Integer,
        "success-count",
        "The number of successes/desired items dropped"
      )
      .min_int_value(1)
    )
  }

  async fn execute(&self, ctx: &Context, interaction: &CommandInteraction, _: &Database) -> Result<()> {
    let probability_input = interaction.get_string("probability").unwrap();
    let try_count = interaction.get_integer("try-count").unwrap() as u64;
    let success_count = interaction.get_integer("success-count").unwrap_or(1) as u64;

    if success_count > try_count {
      let embed = CreateEmbed::new()
        .color(Colors::Red)
        .title("Too many successes")
        .description("`success-count` must be smaller than or equal to `try-count`");

      interaction.reply_embed_ephemeral(&ctx.http, embed).await?;
      return Ok(());
    }

    let probability = match verify_probability(probability_input.as_str()) {
      Some(probability) => probability,
      None => {
        let embed = CreateEmbed::new()
          .color(Colors::Red)
          .title("Invalid format")
          .description("The argument `probability` must be a decimal number or a fraction, and be between 0 and 1.");

        interaction.reply_embed_ephemeral(&ctx.http, embed).await?;
        return Ok(());
      }
    };

    let bin_dist = Binomial::new(try_count as usize, probability as f64);
    // .distribution computes the Cumulative Distribution Function (CDF): P(X <= success_count).
    // To compute P(X >= success_count), we can do 1 - P(X < success_count), which
    // here is written as  or P(X <= success_count - 1)
    let result = 1.0 - bin_dist.distribution((success_count - 1) as f64);
    
    let display_probability = custom_format(result * 100_f64);

    let inner_description = if success_count > 1 {
      format!("**{success_count}** drops (or better)")
    } else {
      "a drop".to_string()
    };

    let try_format = if try_count > 1 { "tries" } else { "try" };

    let embed = CreateEmbed::new()
      .color(Colors::Blue)
      .title("How lucky were you?")
      .description(format!(
        "With a drop chance of **{p}**, getting {m} within **{n}** {t} has a **{x}%** chance of happening.",
        p = probability_input,
        m = inner_description,
        n = try_count,
        t = try_format,
        x = display_probability
      ));

    interaction.reply_embed(&ctx.http, embed).await?;
    Ok(())
  }
}

// Always keep two decimal places (or significant figures, depending on what's relevant)
fn custom_format(input: f64) -> String {
  let leading_digits = input.log10().ceil();
  let trailing_digits = if leading_digits < 0.0 {
    (2.0 - leading_digits).min(8.0) as usize
  } else {
    2
  };

  format!("{:.*}", trailing_digits, input)
}
